### ISSUE TITLE
Read-only floppy image mount option

### ISSUE THREAD

**Issue by bmunger** [dosbox-staging/dosbox-staging#545]
**Is your feature request related to a problem? Please describe.**
No

**Describe the solution you'd like**
Add -ro flag to imgmount to mount read only

**Describe alternatives you've considered**
Setting filesystem permissions to read only is the only alternative.

**Additional context**
None

**Comment by Wengier** [comment]
DOSBox-X allows you to make any disk image read-only by putting a leading colon (:). This allows you to write-protect individual disk images instead of all disk images, e.g.

```imgmount a :disk1.img disk2.img :disk3.img```

The above will write-protect disk1.img and disk3.img, but not disk2.img. This setting can also be turned off in the config file (dosbox-x.conf).

**Comment by dreamer** [comment]
Putting this request in the backlog. Perhaps we should support the same syntax as DOSBox-X, but it's a pretty weird one.

@Wengier why ':' before the file name? Is there any precedent for using such syntax? Outside of Windows, this is a valid character in the filename.

**Comment by kcgen** [comment]
The colon syntax is pointless 'in the know' trickery to save two extra characters versus '-ro', which is explicit and self-documenting for newcomers to dosbox.

Although some games have one or more 'data'  disks and then a 'save game' disk, these same types of games can be installed to HDD (such as the early SSI D&D games); so given this, I would go even further to simplify the interface and just allow a single readonly flag that applies to everything. 

(And wait to see if anyone really does need finer control than that, and assess the need when the time comes).

**Comment by Wengier** [comment]
@dreamer Almost every file name is valid in Linux, so does “-ro” or “-rw”.

The fact that you can specify -ro or -rw more than once makes it non-standard option like the others. On the other hand, the leading colon marker is was made optional in DOSBox-X exactly because it could occasionally be included in the file name. If there is indeed an image file name starting with a colon in Linux (e.g. ":disk.img"), and you do not want to write-protect it, you can temporarily disable the "leading colon write protect image" config option in DOSBox-X, e.g.

```
config -set "dosbox leading colon write protect image=false"
imgmount a :disk.img
config -set "dosbox leading colon write protect image=true"
```

On the other hand, I can understand that -ro or -rw is more explicit and self-documenting for newcomers, especially if you do not need to finer control and want a single readonly flag that applies to everything. Both styles have their advantages of course, and perhaps DOSBox-X can support both for different needs from users.

**Comment by Wengier** [comment]
The latest DOSBox-X code now supports both the read-only marker for individual images and a single -ro for all disk images. By the way, @rderooy did request the ability to write-protect some of the disk images but not the others from the beginning, which made it apparent that a single -ro flag could not solve everything even though it may be useful in many cases.

**Comment by kcgen** [comment]
> By the way, @rderooy did request the ability to write-protect some of the disk images but not the others

Thanks @Wengier. @rderooy, can you recall if it was a hard requirement for a game (or program?) Or if it was more of a personal preference to use floppy images instead of HDD install? 

(Just curious!.. despite having many-fold more dos games today than I did in the BBS days of the 90s, I'm still encountering entirely new use cases!)

**Comment by rderooy** [comment]
@kcgen there are 2 use cases I can think of.
- DOS 4.0x installer (which is pretty buggy) suggests you keep you install disks write-protected, and indeed, I ran into issues that it would corrupt its own install disks if they are not protected. But at the same time it needs one or more disks during install to create a copy of itself. So you need to have some disks protected and not others. And in case your wondering, using multiple drives does not work (even though supposedly it is meant to support it).

- I have run into games in the past (Japanese games) that ran from disk, that required a "user disk" to write to. Obviously your meant to have the game disks protected, and the user disk RW. Don't ask me for names though, it has been a long time.

**Comment by kcgen** [comment]
Fascinating and at the same time hilarious that 4.0 actually corrupts itself!  

I wonder if it was a light weight way to bungle the disks of people who casually copied the OS back in the day; you had one shot at an install, and the disks were hosed on the next attempt sending the person to the store for a legit copy.

**Comment by rderooy** [comment]
@kcgen I don't think so. It would generate too many support calls.

The DOS 4.0x manual clearly states to keep the original disks write protected, so as long as you did that your fine.

I did not test this in detail. But I suspect what is happening, is that during the mandatory disk copy process, if you accidentally swap in the wrong disk it does not notice that and will mess things up.

Unfortunately the disk-swap process in DOSBox is not very user friendly. There is no way for the typical user to see what the currently inserted disk is. Yes there is a message in the LOG as to the currently inserted disk (at least in DOSBox-X, did not check others), but most people do not have the LOG open. So it is easy with a multi disk setup, where you need to constantly swap between disks to end up with the wrong disk image. Some kind of OSD could really help here.

p.s. DOSBox-X has different swap keys for disks and CD's.

**Comment by Wengier** [comment]
@rderooy In the latest DOSBox-X code of my pull request there, IMGMOUNT command without parameters will now show the swap position and total number of swap disks for each drive (and drive number), along with other information. Hope this makes it easier to manage the swap disks.

**Comment by rderooy** [comment]
@Wengier that is a indeed a nice addition. But as you know, once you have started the application in question, such as the DOS 4.0x installer, you cannot run dosbox commands. So the practical usefulness is rather limited. Really you need to have some way of seeing what the currently active image is. Either an OSD, or status bar. Or in the case of DOSBox-X perhaps in the drives menu something like Drives -> A -> Swap Disk: disk2.img

**Comment by dreamer** [comment]
We were thinking about adding it to the title bar (I think I had it even as part of drag'n'drop mounting proof-of-concept build at some point), but after consideration - abandoned that idea (not convenient nor readable enough and not visible in fullscreen). We'll likely implement it as OSD layer instead.

**Comment by Wengier** [comment]
@rderooy Thanks for the suggestion. I have already added a "Drive information" in the "Drive" menu for all drives in the latest DOSBox-X code, so that you can now see information such as the disk type, disk name or mounted path, and swap position even if you are currently running an application.

**Comment by rderooy** [comment]
@Wengier nice! that solves most issues. The only thing missing I can think of is an indicator that the disk image is mounted RO or RW. But we are talking about DOSBox-X in a dosbox-staging feature request :-)

**Comment by dreamer** [comment]
@rderooy thank you for mentioning it; yes, discussion about DOSBox-X features and design decisions is best discussed in DOSBox-X issue tracker, not ours :)

**Comment by Wengier** [comment]
The latest DOSBox-X pull request with the read-only indicator added is now in joncampbell123/dosbox-x#1808, so any further discussions about the feature in DOSBox-X can be done from there, in case @rderooy or anyone else has any further suggestions about it :)

**Comment by Grounded0** [comment]
Fixed by #1509.
